pkg/util/containers/v2/metrics/collector_containerd.go

```diff
@@ -0,0 +1,416 @@
+		log.Debug("Network stats are not available on Windows")
+		return getContainerdNetworkStatsWindows(), nil
+	default:
+		return nil, fmt.Errorf("can't convert the metrics data from container with ID %s", containerID)
+	}
+}
+
+// This method returns interface{} because the metrics could be an instance of
+// v1.Metrics (for Linux) or stats.Statistics (Windows) and they don't share a
+// common interface.
+func (c *containerdCollector) getMetrics(containerID string) (interface{}, error) {
+	container, err := c.client.Container(containerID)
+	if err != nil {
+		return nil, fmt.Errorf("could not get container with ID %s: %s", containerID, err)
+	}
+
+	metricTask, errTask := c.client.TaskMetrics(container)
+	if errTask != nil {
+		return nil, fmt.Errorf("could not get metrics for container with ID %s: %s", containerID, err)
+	}
+
+	metrics, err := typeurl.UnmarshalAny(metricTask.Data)
+	if err != nil {
+		return nil, fmt.Errorf("could not convert the metrics data from container with ID %s: %s", containerID, err)
+	}
+
+	return metrics, nil
+}
+
+func getContainerdStatsLinux(metrics *v1.Metrics, container containers.Container, OCISpec *oci.Spec, processes []containerd.ProcessInfo) *ContainerStats {
+	if metrics == nil {
+		return nil
+	}
+
+	currentTime := time.Now()
+
+	return &ContainerStats{
+		Timestamp: currentTime,
+		CPU:       getContainerdCPUStatsLinux(metrics.CPU, currentTime, container.CreatedAt, OCISpec),
+		Memory:    getContainerdMemoryStatsLinux(metrics.Memory),
+		IO:        getContainerdIOStatsLinux(metrics.Blkio, processes),
+	}
+}
+
+func getContainerdCPUStatsLinux(cpuStat *v1.CPUStat, currentTime time.Time, startTime time.Time, OCISpec *oci.Spec) *ContainerCPUStats {
+	if cpuStat == nil {
+		return nil
+	}
+
+	res := ContainerCPUStats{}
+
+	if cpuStat.Usage != nil {
+		res.Total = uint64ToFloat64Ptr(cpuStat.Usage.Total)
+		res.System = uint64ToFloat64Ptr(cpuStat.Usage.Kernel)
+		res.User = uint64ToFloat64Ptr(cpuStat.Usage.User)
+	}
+
+	if cpuStat.Throttling != nil {
+		res.ThrottledPeriods = uint64ToFloat64Ptr(cpuStat.Throttling.ThrottledPeriods)
+		res.ThrottledTime = uint64ToFloat64Ptr(cpuStat.Throttling.ThrottledTime)
+	}
+
+	res.Limit = getContainerdCPULimit(currentTime, startTime, OCISpec)
+
+	return &res
+}
+
+func getContainerdMemoryStatsLinux(memStat *v1.MemoryStat) *ContainerMemStats {
+	if memStat == nil {
+		return nil
+	}
+
+	res := ContainerMemStats{
+		RSS:   uint64ToFloat64Ptr(memStat.RSS),
+		Cache: uint64ToFloat64Ptr(memStat.Cache),
+	}
+
+	if memStat.Usage != nil {
+		res.UsageTotal = uint64ToFloat64Ptr(memStat.Usage.Usage)
+		res.Limit = uint64ToFloat64Ptr(memStat.Usage.Limit)
+	}
+
+	if memStat.Kernel != nil {
+		res.KernelMemory = uint64ToFloat64Ptr(memStat.Kernel.Usage)
+	}
+
+	if memStat.Swap != nil {
+		res.Swap = uint64ToFloat64Ptr(memStat.Swap.Usage)
+	}
+
+	return &res
+}
+
+func getContainerdIOStatsLinux(blkioStat *v1.BlkIOStat, processes []containerd.ProcessInfo) *ContainerIOStats {
```

Comment:
This function is very similar to the one in the Docker collector (https://github.com/DataDog/datadog-agent/pull/9612). We should probably extract this so it can be called from multiple collectors.